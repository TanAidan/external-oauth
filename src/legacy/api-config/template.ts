import at from 'lodash.at'

import { CustomError } from '../errors'

// Matches escaped characters eg. "\$" or variables with optional surrounding whitespace eg. "${  name  }"
const parserRegex = /\\.|\$\{\s*([^}\s]*)\s*\}/g

const expandString = (template: string, templatePath: string, variables: any) =>
  template.replace(parserRegex, (match: string, variableName: string) => {
    if (match[0] === '\\') {
      return match[1]
    }

    const value = at<string | undefined>(variables, variableName)[0]

    if (value === undefined) {
      throw new UndefinedVariable(templatePath, variableName)
    }

    return value
  })

const expandArray = (template: any, templatePath: string, variables: any) =>
  template.map((element, i: number) => expand(element, `${templatePath}[${i}]`, variables))

const expandObject = (template: any, templatePath: string, variables: any) => {
  const result = {}

  for (const [attributeName, childTemplate] of Object.entries(template)) {
    const childPath = templatePath ? `${templatePath}.${attributeName}` : attributeName

    result[attributeName] = expand(childTemplate, childPath, variables)
  }

  return result
}

const expand = (template: any, templatePath: string, variables: any) => {
  if (Array.isArray(template)) {
    return expandArray(template, templatePath, variables)
  }

  if (typeof template === 'string') {
    return expandString(template, templatePath, variables)
  }

  if (typeof template === 'number') {
    return template
  }

  return expandObject(template, templatePath, variables)
}

/**
 * Performs variable interpolation on strings. The strings may be contained within
 * nested objects or arrays. Numbers are returned unchanged.
 *
 * Examples variables `{ name: 'Bearer', address: { city: 'Paris' } }`:
 *
 * Example cases:
 *
 *   `{ greeting: 'Hi ${name}', count: 5 }` => `{ greeting: 'Hi Bearer', count: 5 }`
 *
 *   `['City is ${address.city}', { busy: true }]` => `['City is Paris', { busy: true }]`
 *
 *   `'Hey there "${name}"'` => `'Hey there "Bearer"'`
 *
 *   `'Not a variable: \\${name}'` => `'Not a variable: ${name}'`
 *
 *   `{ a: 'Oops ${some.age}' }` => throws `UndefinedVariable(templatePath: 'a', variableName: 'some.age')`
 */
export const expandTemplate = (template: any, variables: any) => expand(template, '', variables)

export class UndefinedVariable extends CustomError {
  constructor(public templatePath: string, public variableName: string) {
    super(`Undefined variable ${variableName} used in template at ${templatePath}`)
  }
}
