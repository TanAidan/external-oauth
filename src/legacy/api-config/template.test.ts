import { expandTemplate } from './template'

describe('expandTemplate', () => {
  it('expands variables in an object', () => {
    const result = expandTemplate({ a: 'hello ${x.who}', b: { c: '${y}' } }, { x: { who: 'world' }, y: 1 })

    expect(result).toEqual({ a: 'hello world', b: { c: '1' } })
  })

  it('expands variables in a string', () => {
    const result = expandTemplate('hello ${x.who}', { x: { who: 'world' } })

    expect(result).toEqual('hello world')
  })

  it('expands variables in an array', () => {
    const result = expandTemplate(['hello ${x.who}', 'hey!'], { x: { who: 'world' } })

    expect(result).toEqual(['hello world', 'hey!'])
  })

  it('returns numbers as-is', () => {
    const result = expandTemplate(42, {})

    expect(result).toEqual(42)
  })

  it('does NOT expand escaped variables', () => {
    const result = expandTemplate('hello \\${x.who}', { x: { who: 'world' } })

    expect(result).toEqual('hello ${x.who}')
  })

  it('allows the escape character to be used when escaped', () => {
    const result = expandTemplate('hello\\\\there', {})

    expect(result).toEqual('hello\\there')
  })

  describe('when NO value was supplied for a variable', () => {
    it('raises an UndefinedVariable error with the correct path for objects', () => {
      expect(() => expandTemplate({ a: { b: '${x.who}' } }, {})).toThrowErrorMatchingSnapshot()
    })

    it('raises an UndefinedVariable error with the correct path for arrays', () => {
      expect(() => expandTemplate({ a: ['hey', '${x.who}'] }, {})).toThrowErrorMatchingSnapshot()
    })
  })
})
