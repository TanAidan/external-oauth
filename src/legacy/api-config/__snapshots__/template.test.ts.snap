// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expandTemplate when NO value was supplied for a variable raises an UndefinedVariable error with the correct path for arrays 1`] = `"Undefined variable x.who used in template at a[1]"`;

exports[`expandTemplate when NO value was supplied for a variable raises an UndefinedVariable error with the correct path for objects 1`] = `"Undefined variable x.who used in template at a.b"`;
